Support BigInteger in GraphQL (#744 and #909)

echauchot opened the pull request:
This PR covers both #744 and #909. Here is what it does:

- Registers extended graphQL scalars /BigInteger in the server (always, no conf item): reused and renamed unused addlTypes to extendedScalarTYpes
- Add a new **extendedScalarTypes** configuration in package.json (in the compiler configuration) which defaults to false for backward compatibility. So the default package.json does not change
- Registers extended graphQL scalars /BigInteger in the compiler based on the package.json conf item
- Reference new extendedScalarTypes in the specs doc
- map all Calcite BIGINT types to ExtendedScalars.GraphQLBigInteger based on the compiler config parameter
- Add a new usecase test with BigInteger support enabled: **simple-with-bigint-support** that does a simple select on the ecommerce/Orders  table with BIGINT as primary key
- Tested with both graphQL schema provided and not provided on UseCaseCompileTest and committed the corresponding snapshot 
- Tested with both graphQL schema provided and not provided with test and run goals on FullUsecasesIT and committed the corresponding snapshot 

@henneberger @mbroecheler PTAL

henneberger reviewed (approved):
Looks good. One comment about naming.

echauchot commented:
> Looks good. One comment about naming.

@henneberger thanks for your review. I don't see your naming comment, have you published your review ?

echauchot commented:
What is strange regarding the build errors is that **UseCaseCompileTest** runs fine when run alone but part of mvn verify, some tests start to fail:
- **simple-with-bigint-support**: fails with a snapshot mistmatch. It is expecting a  graphql schema with BigInteger but the generated schema still uses Float as if the use case was using the default package.json (with extendedScalarTypes disabled) instead of the packaged one.
- **seedshop-avro**: fails with _Not a valid SQRL data type_ in TableConverter.sourceToTable() with both provided package.json

henneberger commented on sqrl-calcite/src/main/java/com/datasqrl/calcite/QueryPlanner.java:
```diff
@@ -203,9 +203,15 @@ public RelDataType parseDatatype(String datatype) {
     }
 
+    if (datatype.equalsIgnoreCase("BigInteger")) {
+      return this.cluster.getTypeFactory().createSqlType(SqlTypeName.BIGINT);
```
We could use the graphql-java base set of scalars (`scalar GraphQLBigInteger`) so all of the types can be mapped uniformly. https://github.com/graphql-java/graphql-java-extended-scalars?tab=readme-ov-file#java-primitives

echauchot replied:
What do you mean ?
`scalar GraphQLBigInteger` maps to `"BigInteger"`  String litteral in [JavaPrimitives](https://github.com/graphql-java/graphql-java-extended-scalars/blob/9e283115657220fd191e7a03d9ecf07964746aed/src/main/java/graphql/scalars/java/JavaPrimitives.java#L373)

echauchot replied:
I pushed comments and test related to that

henneberger replied:
Calling it BigInteger may be confusing since there is also a "BigInt" that is a java Long type. The Numeric type is the type that maps to java BigInteger. Since we have one type system for both sql and graphql, it may be better to disambiguate this by calling it "GraphQLBigInteger"

echauchot replied:
Thanks it is clearer. I agree. I pushed a fix. PTAL

echauchot commented:
> What is strange regarding the build errors is that **UseCaseCompileTest** runs fine when run alone but part of mvn verify, some tests start to fail:
> 
> * **simple-with-bigint-support**: fails with a snapshot mistmatch. It is expecting a  graphql schema with BigInteger but the generated schema still uses Float as if the use case was using the default package.json (with extendedScalarTypes disabled) instead of the packaged one.
> * **seedshop-avro**: fails with _Not a valid SQRL data type_ in TableConverter.sourceToTable() with both provided package.json

My bad, my local m2 repo was somehow out of sync. With a full maven build, there is no error except indeed the timestamp mismatch in FullUseCasesIT: the new use case generated a snapshot with my local timezone which is different from the timezone the git servers are in.

We should make the tests timezone agnostic so that the locally generated snapshots are comparable to the expected ones in the test infrastructure. I suggest programmatically forcing the timezone in the tests for both snapshot generation and comparison. WDYT ?

In the meantime, I'll update the snapshot for the new test to the expected one.

echauchot commented:
> > What is strange regarding the build errors is that **UseCaseCompileTest** runs fine when run alone but part of mvn verify, some tests start to fail:
> > 
> > * **simple-with-bigint-support**: fails with a snapshot mistmatch. It is expecting a  graphql schema with BigInteger but the generated schema still uses Float as if the use case was using the default package.json (with extendedScalarTypes disabled) instead of the packaged one.
> > * **seedshop-avro**: fails with _Not a valid SQRL data type_ in TableConverter.sourceToTable() with both provided package.json
> 
> My bad, my local m2 repo was somehow out of sync. With a full maven build, there is no error except indeed the timestamp mismatch in FullUseCasesIT: the new use case generated a snapshot with my local timezone which is different from the timezone the git servers are in.
> 
> We should make the tests timezone agnostic so that the locally generated snapshots are comparable to the expected ones in the test infrastructure. I suggest programmatically forcing the timezone in the tests for both snapshot generation and comparison. So I updated the test to do that.
> 
> Now the timestamps in all the snapshots are exactly the ones that are inserted in the synthetic source. Before that there was a difference that might have been a bug. So I updated all the snapshots for all the use cases with the timezone forcing.

As discussed, I'm no more updating all the snapshots and trying to fix the timezone issue as it is part of a broader timezone plan.  I replaced BigIntOrdersTest.snapshot with the snapshot generated in the CI (with UTC-8 timezone) to workaround this local timezone issue.

Regarding a ticket opening for fixing timezone issue, I found [#895](https://github.com/DataSQRL/sqrl/issues/895) which mentions setting a fixed timezone to UTC in the tests. This tickets seems a good fit to me so I just added a comment to it to trace the observed behavior in this PR.

echauchot commented:
signed all the commits and pushed

echauchot commented:
> signed all the commits and pushed

But when we hit github button rebase and merge, the signatures seem to be gone :smile:

echauchot commented:
@henneberger thanks for your review
